umap/static/umap/js/modules/sync/engine.js

```diff
@@ -27,39 +69,201 @@ export class SyncEngine {
     this.transport = undefined
   }
 
+  upsert(subject, metadata, value) {
+    this._send({ verb: 'upsert', subject, metadata, value })
+  }
+
+  update(subject, metadata, key, value) {
+    this._send({ verb: 'update', subject, metadata, key, value })
+  }
+
+  delete(subject, metadata, key) {
+    this._send({ verb: 'delete', subject, metadata, key })
+  }
+
+  _send(inputMessage) {
+    let message = this._operations.addLocal(inputMessage)
+
+    if (this.offline) return
+    if (this.transport) {
+      this.transport.send('OperationMessage', message)
+    }
+  }
+
   _getUpdater(subject, metadata) {
     if (Object.keys(this.updaters).includes(subject)) {
       return this.updaters[subject]
     }
     throw new Error(`Unknown updater ${subject}, ${metadata}`)
   }
 
-  // This method is called by the transport layer on new messages
+  _applyOperation(operation) {
+    const updater = this._getUpdater(operation.subject, operation.metadata)
+    updater.applyMessage(operation)
+  }
+
+  _updateLastKnownHLC(remoteHLC) {
+    if (!this._lastKnownHLC || remoteHLC > this._lastKnownHLC) {
+      this._lastKnownHLC = remoteHLC
```

Comment:
I don't get why we have both `operation._hlc` and `engine._lastKnownHLC`, are those the same value at the end ?

Reply:
I've refactored a bit the code in commit 3089da160a9d303f8526cb4ea05908271d071e97 to be clearer and easier to follow, effectively removing this from the main `SyncEngine` class, and putting it in the `Operations` class instead.